eth/db_upgrade.go

```diff
@@ -0,0 +1,343 @@
+			glog.V(logger.Info).Infof("Database conversion successful")
+			db.Put(useSequentialKeys, []byte{42})
+		}
+		if err != nil {
+			glog.V(logger.Error).Infof("Database conversion failed: %v", err)
+		}
+		close(stoppedChn)
+	}()
+
+	return func() {
+		close(stopChn)
+		<-stoppedChn
+	}
+}
+
+// upgradeSequentialCanonicalNumbers reads all old format canonical numbers from
+// the database, writes them in new format and deletes the old ones if successful.
+func upgradeSequentialCanonicalNumbers(db ethdb.Database, stopFn func() bool) (error, bool) {
+	prefix := []byte("block-num-")
+	it := db.(*ethdb.LDBDatabase).NewIterator()
+	it.Seek(prefix)
+	cnt := 0
+	for bytes.HasPrefix(it.Key(), prefix) {
+		keyPtr := it.Key()
+		if len(keyPtr) < 20 {
+			cnt++
+			if cnt%100000 == 0 {
+				glog.V(logger.Info).Infof("converting %d canonical numbers...", cnt)
+			}
+			number := big.NewInt(0).SetBytes(keyPtr[10:]).Uint64()
+			newKey := []byte("h12345678n")
+			binary.BigEndian.PutUint64(newKey[1:9], number)
+			if err := db.Put(newKey, it.Value()); err != nil {
+				return err, false
+			}
+			if err := db.Delete(keyPtr); err != nil {
+				return err, false
+			}
+		}
+
+		if stopFn() {
+			return nil, true
+		}
+		it.Next()
+	}
+	if cnt > 0 {
+		glog.V(logger.Info).Infof("converted %d canonical numbers...", cnt)
+	}
+	return nil, false
+}
+
+// upgradeSequentialBlocks reads all old format block headers, bodies, TDs and block
+// receipts from the database, writes them in new format and deletes the old ones
+// if successful.
+func upgradeSequentialBlocks(db ethdb.Database, stopFn func() bool) (error, bool) {
+	prefix := []byte("block-")
+	it := db.(*ethdb.LDBDatabase).NewIterator()
+	it.Seek(prefix)
+	cnt := 0
+	for bytes.HasPrefix(it.Key(), prefix) {
+		keyPtr := it.Key()
+		if len(keyPtr) >= 38 {
+			cnt++
+			if cnt%10000 == 0 {
+				glog.V(logger.Info).Infof("converting %d blocks...", cnt)
+			}
+			// convert header, body, td and block receipts
+			var keyPrefix [38]byte
+			copy(keyPrefix[:], keyPtr[0:38])
+			hash := keyPrefix[6:38]
+			if err := upgradeSequentialBlockData(db, hash); err != nil {
+				return err, false
+			}
+			// delete old db entries belonging to this hash
+			for bytes.HasPrefix(it.Key(), keyPrefix[:]) {
+				if err := db.Delete(it.Key()); err != nil {
+					return err, false
+				}
+				it.Next()
+			}
+			if err := db.Delete(append([]byte("receipts-block-"), hash...)); err != nil {
+				return err, false
+			}
+		} else {
+			it.Next()
+		}
+
+		if stopFn() {
+			return nil, true
+		}
+	}
+	if cnt > 0 {
+		glog.V(logger.Info).Infof("converted %d blocks...", cnt)
+	}
+	return nil, false
+}
+
+// upgradeSequentialOrphanedReceipts removes any old format block receipts from the
+// database that did not have a corresponding block
+func upgradeSequentialOrphanedReceipts(db ethdb.Database, stopFn func() bool) (error, bool) {
+	prefix := []byte("receipts-block-")
+	it := db.(*ethdb.LDBDatabase).NewIterator()
```

Comment:
I think this should be re-written partially on the `LDBDatabase` and use [this commit](https://github.com/ethereum/go-ethereum/pull/2618/commits/d2d4bcfad57b2c2d4003f9d8a0f10c5862f0f63c) so that we can also use `leveldb` iterators.

This will help us because it will allow us to use ranges for the iterator:

```go
iter := db.NewIterator(&util.Range{Start: prefix, Limit: nil}, nil)
for iter.Next() {
    // ...
}
```

This might actually be even faster too since you don't have to check each key manually using `bytes.HasPrefix`.